pbs_benchpress --use-current-setup needs manager and operator privileges

sujatapatnaik52 opened the pull request:
#### Describe Bug or Feature
`pbs_benchpress` has an option `--use-current-setup` which runs a testsuite with the current state of PBS. After doing a fresh installation of PBS, when `pbs_benchpress` is run with --use-current-setup, an error **Unauthorized request** is observed while running a `qmgr` command in the test case. 

#### Describe Your Change
Since the test execution user needs manager privileges for creating resources in PBS Server, Added the test execution user as manager role at the testsuite level. Giving `MGR_USER` and `MGR_OPER` as manager and operator roles respectively similar to the default roles assigned without `--use-current-setup`

#### Attach Test and Valgrind Logs/Output
[after_fix.txt](https://github.com/PBSPro/pbspro/files/4243013/after_fix.txt)
[before_fix.txt](https://github.com/PBSPro/pbspro/files/4243014/before_fix.txt)

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

hirenvadalia commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -440,6 +440,8 @@ class PBSTestSuite(unittest.TestCase):
     moms = None
     comms = None
+    users = ""
+    added_users = []
```
I don't think you need 2 separate variables to store a username and its role. It can be done by one variable also.

hirenvadalia replied:
Also, I suggest renaming `added_users` to something like `added_mgrs_oprs` for more readability

hirenvadalia commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -450,6 +452,19 @@ def setUpClass(cls):
+                if role not in server_stat or user not in server_stat[role]:
+                    attr = {role: (INCR, user)}
+                    cls.server.manager(MGR_CMD_SET, SERVER, attr, sudo=True)
+                    cls.added_users.append(user)
```
rather than adding just user in `added_users`, you can add user + role, so you don't need 2nd class variable.

hirenvadalia commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -450,6 +452,19 @@ def setUpClass(cls):
+            for user, role in cls.users.items():
+                if role not in server_stat or user not in server_stat[role]:
+                    attr = {role: (INCR, user)}
+                    cls.server.manager(MGR_CMD_SET, SERVER, attr, sudo=True)
```
Also, these added mgrs and oprs will be saved in setup (which is not part of the original setup which user asked to save and use) so shouldn't we need to take care of removing added mgrs and oprs while restoring saved setup?

sujatapatnaik52 replied:
I agree with you. If I remove the added mgrs and oprs in teardown(while restoring the saved setup), I may have to add it again in the next testcase's setup. so even though I have added it in setupclass I will have to add it in setup as well. Does that sound fine?

hirenvadalia replied:
On second thought (as per our offline skype chat): I think we should move this setting mgrs and oprs into `save_configuration` and reverting it to `load_configuration`, that way if someone else other than `setUp/tearDown(class)` is using save/load configuration then they will be also beneficial for this change.

sujatapatnaik52 replied:
Thank you @hirenvadalia for the suggestion. I am making the change.

sujatapatnaik52 replied:
@hirenvadalia As per your suggestion, I am using only one variable for the fix. Please look at it.

Also, I tried  adding the managers and operators in `save_configuration` and `load configuration`. It works upto the point where `load_configuration` is called in teardown. I am removing the users which were added in setup. But the catch is `load_configuration` is also called in `teardownclass`, by the time, it reaches there, the users are already removed in `teardown` and because of which there is **Unauthorized request** in `teardownclass`. 

As discussed offline, I feel `save_configuration` and `load_configuration` are API's as per their names i.e. related to configuration changes. Adding users is kind of a pre-requisite which we are doing based on our requirement.

hirenvadalia replied:
" Adding users is kind of a pre-requisite."
Well, I don't agree on this as this is not pre-requisite, those mgrs and oprs are needed by save/load configuration only... so we should set and reset in those methods only rather than asking the user to do before using those methods. I understand there will be overhead to set and reset mgrs and oprs but we will get save/load config method fully independent of anything else, and a user can use without any headache.

sujatapatnaik52 replied:
I may have to add managers and operators in teardownclass before load_configuration and remove them again after the configuration is loaded. This seems to me as a bit of overhead. 

Please correct me if I am wrong. With my current changes, the test execution user is added as a manager role in the `setupclass` of `PBSTestSuite` if used with `--use-current-setup`. If the user is using `save/load configuration` in the test case, then the user shall still be permitted to use the API's.

hirenvadalia replied:
"Please correct me if I am wrong. With my current changes, the test execution user is added as a manager role in the setupclass of PBSTestSuite if used with --use-current-setup. If the user is using save/load configuration in the test case, then the user shall still be permitted to use the API's."

Yes anybody (within PTL fw + tests) can use save and load conf method. and that's why I am suggesting that we should make those methods independent of anything.

I understand there will be overhead, but everything comes with a price :smile: 

Or try to modify save and load conf methods such a way that it doesn't need mgrs and opers set? like use sudo or root user when doing privilege operation inside save/load conf methods

sujatapatnaik52 replied:
okay how about this - Let's make load_configuration and save_configuration work with sudo privileges. The managers and operators will still be needed for the test execution to work properly so that can be added independent of these API's maybe after the configuration is saved. Sounds good?

sujatapatnaik52 commented:
@hirenvadalia The `save_configuration` and `load_configuration` has most of the commands which run with sudo privileges, It is just the other API's like `delete_current_state` which needed the same. I added a optional sudo parameter to those API's. I have added the managers and operators in setup and removing them in teardown. Please look at the latest commit.

hirenvadalia commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -483,7 +484,28 @@ def setUp(self):
-            self.server.delete_nodes()
+            self.server.delete_nodes(True)
+            attr = {}
+            test_user = pwd.getpwuid(os.getuid())[0] + '@*'
```
Now when you are using sudo everywhere why do we need to set mgrs and opers?

sujatapatnaik52 replied:
The `revert_moms` is failing. Also, In the test case, the user may create vnodes, queues or do some other manager oriented operations. So I feel adding them will be the right thing to do.

hirenvadalia replied:
"Also, In the test case, the user may create vnodes, queues or do some other manager oriented operations"
This test case has to take care of it, not setUp.

"The revert_moms is failing."
I suggest using sudo there also... We should keep one way of doing stuff, either sudo or with mgrs/opers... not both of it.

sujatapatnaik52 commented:
Thank you @hirenvadalia for the thorough review and suggestions. Requesting one more review from @pbspro-maintainers.

sujatapatnaik52 commented:
[smoketest_run_logs_with_use_current_setup](https://github.com/PBSPro/pbspro/files/4244355/smoketest_logs_use_current_setup.txt)

hirenvadalia commented:
> should PTL really be doing this? --use-current-setup kind of relies on the test runner to set things up such that the test runs properly right? Also, with --use-current-setup, PTL shouldn't be creating any new resources/nodes/queues etc., so why does it need operator and manager privileges ?

I also agree @agrawalravi90 with you here but the problem is, in PTL (from starting) we set the current user as mgr/oper, and now we have many tests written keeping that in mind so either we have to change all those tests not to assume anything or adopt the same env in `--use-current-setup`.
And easiest is to adopt env in `--user-current-setup` rather than changing hundreds of test cases.

Also, we can ask the user to make sure it adds the current user as mgr/opers but that will be a headache for the user to keep this mind and do it every time before running pbs_benchpress.
So from a usability point of view, we thought changing fw will be good.

HTH!

agrawalravi90 commented:
Thanks for explaining Hiren. Ok, I agree that this sounds like the most sensible thing to do right now. But we should document this. Can we modify the doc:
https://pbspro.atlassian.net/wiki/spaces/PD/pages/557088789/Design+for+a+supported+way+to+change+default+setup+in+PTL

to mention this as a caveat ?

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5296,7 +5296,7 @@ def save_configuration(self, outfile, mode='a'):
             conf['qmgr_print_sched'] = ret['out']
         ret = self.du.run_cmd(self.hostname, [pbsnodes, '-av'],
-                              logerr=False, level=logging.DEBUG)
+                              logerr=False, level=logging.DEBUG, sudo=True)
```
why does this need a sudo ? it's just pbsnodes -av command right?

sujatapatnaik52 replied:
`pbsnodes -av` without `sudo` i.e. if this command is run as a test execution user, it does not list all the nodes attributes like Port, version etc. We just proceed with saving the configuration without `Port`. While loading configuration, we create the node using `Port` and that gives a `KeyError`. So, I had to add this.

agrawalravi90 replied:
thanks for explaining, can you please add this as a comment in the code so that someone like me doesn't remove it thinking it's unnecessary ? :)

sujatapatnaik52 replied:
Added the comment. Also, I have updated the design document. Please review both.

bhroam replied:
As a note for the future, you didn't need to use sudo.  There is no difference between running as root and another manager.  You could have done a runas=MGR_USR and it would have been a bit more secure.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -458,6 +458,7 @@ def setUpClass(cls):
             else:
                 cls.logger.error("Failed to save custom setup")
                 raise Exception("Failed to save custom setup")
+            cls().add_mgrs_opers(cls.server)
```
this looks kind of hack-y .. you are creating an instance of the class via cls() to call a non-classmethod from a classmethod. why not just make add_mgrs_opers() a classmethod? Then you can just call it as "cls.add_mgrs_opers()" here. Inside setUp() for the normal code path, you can just call it as self.add_mgrs_opers(). You'll have to change "server" to "cls.server" inside the method

sujatapatnaik52 replied:
I tried this yesterday and I was getting some error so I did not debug further. Maybe I was surely doing something wrong. I have made the changes as per your suggestion. Please review the latest commit.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1308,6 +1309,38 @@ def revert_moms(self, force=False):
+                if len(add_users) > 0:
+                    attr[role] = (INCR, ",".join(add_users))
+        if len(attr) > 0:
+            self.server.manager(MGR_CMD_SET, SERVER, attr, sudo=True)
```
you are passing in 'server' object so you don't need to call it as "self.server". If you take my suggestion for converting this into a class method, then you'll have to use 'cls.server' everywhere in the method

sujatapatnaik52 commented:
Thank you @agrawalravi90 and @hirenvadalia for the sign off.
